addons/binding/org.openhab.binding.icloud/ESH-INF/thing/thing-types.xml

```diff
@@ -0,0 +1,138 @@
+			<channel id="numberOfDevices" typeId="NumberOfDevices" />
+			<channel id="refresh" typeId="Refresh" />
+		</channels>
+
+		<config-description>
+			<parameter name="appleId" type="text" required="true">
+				<context>email</context>
+				<label>Apple Id</label>
+				<description>Apple Id (e-mail) to access iCloud information.</description>
+			</parameter>
+			<parameter name="password" type="text" required="true">
+				<context>password</context>
+				<label>Password</label>
+				<description>Apple iCloud password for the given Id.</description>
+			</parameter>
+			<parameter name="refreshTimeInMinutes" type="integer" min="5" max="65535" unit="min">
+				<label>Refresh</label>
+				<description>Refresh time in minutes.</description>
+				<default>5</default>
+			</parameter>
+			<parameter name="googleAPIKey" type="text">
+				<label>Google API Key</label>
+				<description>Google API key to be used for address lookup calls (to translate coordinates into a human readable string). Without a key the number of daily request are limited and the bindings address lookup functionality will stop working.</description>
+			</parameter>
+		</config-description>
+	</bridge-type>
+
+	<thing-type id="device">
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="bridge" />
+		</supported-bridge-type-refs>
+
+		<label>iCloud Device</label>
+
+		<channels>
+			<channel typeId="BatteryStatus" id="batteryStatus" />
+			<channel typeId="system.battery-level" id="batteryLevel" />
+			<channel typeId="FindMyPhone" id="findMyPhone" />
+			<channel typeId="Location" id="location" />
+			<channel typeId="LocationAccuracy" id="locationAccuracy" />
+			<channel typeId="LocationLastUpdate" id="locationLastUpdate" />
+			<channel typeId="DistanceFromHome" id="distanceFromHome" />
+			<channel typeId="AddressStreet" id="addressStreet" />
+			<channel typeId="AddressCity" id="addressCity" />
+			<channel typeId="AddressCountry" id="addressCountry" />
+			<channel typeId="FormattedAddress" id="formattedAddress" />
+			<channel typeId="DeviceName" id="deviceName" />
+		</channels>
+
+		<representation-property>deviceId</representation-property>
+
+		<config-description>
+			<parameter name="deviceId" type="text" required="true">
+				<label>Device id</label>
+			</parameter>
+		</config-description>
+
+	</thing-type>
+
+	<channel-type id="NumberOfDevices">
+		<item-type>Number</item-type>
+		<label>Number of devices</label>
+		<description>Number of registered devices with this iCloud account.</description>
+	</channel-type>
+	<channel-type id="Owner">
+		<item-type>String</item-type>
+		<label>Owner</label>
+		<description>Registered owner of this iCloud account.</description>
+	</channel-type>
+	<channel-type id="Refresh">
+		<item-type>Switch</item-type>
+		<label>Refresh</label>
+		<description>Update all devices registered with this account immediately.</description>
+	</channel-type>
+	<channel-type id="EnableAddressLookup">

```

Comment:
This must not be a channel, but rather a binding configuration parameter.

Reply:
But I would want to question this feature in general: This does not seem to be any feature that is specific to this binding - it is a general translation from geocoordinates to an address. As such, it does not belong in here, but we should think about a proper way to make this feature available to ALL bindings. It looks to me as a nice candidate for a [profile](https://www.eclipse.org/smarthome/documentation/concepts/profiles.html).

Reply:
sounds like something to consider after the PR is merged!?

Reply:
Yes, but also to be removed here before the merge.

Reply:
Remove the channel but leave the lookup I hope. It is incredibly useful. I agree it would be amazing to outsource the lookup function as a profile or imho also as a Transformation, after the PR was merged for v2.2

Reply:
I´ve removed the unused channel type (was a left over); the feature I currently kept (will be enabled if a google API key is defined). Let me know if I need to remove it to get the binding merged.

Reply:
> Remove the channel but leave the lookup I hope.

No, I was talking about the whole feature. Better let us start with something small, but stable instead of introducing a feature now that we will remove from the binding in the next release, because there is another mechanism by then. Breaking the binding for users is nothing we should aim for if we can foresee it right now already.

> It is incredibly useful.

If so, it must be incredibly useful for all people that currently use OwnTracks or other location services - and they managed to live without it for now as well ;-)

I agree that it is a neat feature, but that's also why I plead for helping working with me to get this smoothly integrated for the 2.3 release.

Reply:
Ok ... I´ll remove the address lookup and let you know once it´s done.

Reply:
Done

Reply:
A million thanks for your understanding - I owe you something 🙇

Reply:
FTR, here's an issue to continue the discussion: https://github.com/eclipse/smarthome/issues/4748